### PR TITLE
Changes to onChange target and type definitions

### LINKED ISSUE #3173: Select event.value !== event.target.value
When I change the option in a Select, `event.value` is the one I selected, and `event.target.value` is the previously selected option. This is obviously not the correct behavior.

This is a pretty simple issue so unless this is unknown behavior / hard to repro, I'll leave it at that.

### Your Environment

- Grommet version: 2.7.1
- Browser Name and version: Chrome 74.0.3729.169
- Operating System and version (desktop or mobile): Desktop

### PR DESCRIPTION
**Pull request by jimenezff**
#### What does this PR do?
- Keep onChange original target and updated value. Fixes #3173 and #4668
- Updated type definition to reflect this change, among other changes.
- Updated ts story to correctly use the new type definition.
- Added new tests to cover some of the changes to Select.js and new type definitions.

#### Where should the reviewer start?
Changes to `Select.js`
 
#### What testing has been done on this PR?
The added tests to `Select-test.js`

#### How should this be manually tested?
Could be by running tests `event.target.value === event.target` and `onChange target isn't undefined when using valueLabel` with master branch (they will fail)

#### What are the relevant issues?
 #3173  #4668  #3165

#### Do the grommet docs need to be updated?
no
#### Should this PR be mentioned in the release notes?
no
#### Is this change backwards compatible or is it a breaking change?
might be an issue if some people was using onChange's `event.target.value` as a way to obtain the previous value (see #3173 for more details). However, that is not the expected behavior and it wasn't documented as such.

### CODE REVIEW

**Review by jcfilben — COMMENTED** [review]
When testing with some of the storybook stories (Controlled and CustomValueAndIcon) I got the following error "Argument of type 'SelectOption' is not assignable to parameter of type 'SetStateAction<string>'. Type 'number' is not assignable to type 'SetStateAction<string>'."

**Comment by jimenezff** [comment]
Thanks! I'll take a look later today

**Comment by jimenezff** [comment]
@jcfilben what do you think about this implementation with generics?

- Also, `selected` prop (also in `onChange`'s event) is deprecated according to the docs. So, what do you think about removing them or adding a JSDoc `@deprecated` tag?

- This story `src\js\components\Select\stories\Seasons.js` makes use of selected. I adapted it to use `value` instead, and added it to the typescript stories.

File: src/js/components/Select/stories/typescript/Controlled.tsx
Comment (jcfilben) [inline src/js/components/Select/stories/typescript/Controlled.tsx:31]: Story should use new storybook story structure (See #4651)

File: src/js/components/Select/stories/typescript/CustomValueAndIcon.tsx
Comment (jcfilben) [inline src/js/components/Select/stories/typescript/CustomValueAndIcon.tsx:46]: This story should also use new storybook story structure

File: src/js/components/Select/stories/typescript/Seasons.tsx
Comment (jcfilben) [inline src/js/components/Select/stories/typescript/Seasons.tsx:110]: This story should also use new storybook story structure

**Review by jcfilben — COMMENTED** [review]
Nice job switching the Seasons story to use `value` instead of `selected`. I think a comment could be added to the Select/index.d.ts story about `selected` being deprecated but I don't think we should completely remove `selected` because of backwards compatibility issues.

**Review by ShimiSun — COMMENTED** [review]
I think we should avoid adding more stories to TS issues unless they provide additional functional value that other stories don't, it will be helpful if you can attach a snippet of the code (story) you were testing with instead of adding it to the PR.

**Comment by jimenezff** [comment]
Sure I can remove the new ones. Two of them are simply rewrites of existing js ones. However, maybe we could keep the new `Seasons.tsx` story which is making use of `value` instead of the deprecated `selected`. If so, should I delete the old `Seasons.js`?

**Comment by ShimiSun** [comment]
> Sure I can remove the new ones. Two of them are simply rewrites of existing js ones. However, maybe we could keep the new Seasons.tsx story which is making use of value instead of the deprecated selected. If so, should I delete the old Seasons.js?

Good comment. Let's remove the deprecated usage and keep the more recommended usage story.

File: src/js/components/Select/index.d.ts
Comment (jcfilben) [inline src/js/components/Select/index.d.ts:73]:
```suggestion
  value?: ValueType;
```
Value should have `?` to support uncontrolled forms (see Uncontrolled Select story)

**Comment by ShimiSun** [comment]
@jcfilben any other feedback on this PR?

**Review by jcfilben — APPROVED** [review]
No other feedback, this PR looks good to me

File: src/js/components/Select/Select.js
Comment (ShimiSun) [inline src/js/components/Select/Select.js:144]: I suspect this will be a breaking change.

Comment (jcfilben) [reply src/js/components/Select/Select.js:144]: Yeah it would be a breaking change, here is what was said about it in the PR description:

> might be an issue if some people was using onChange's event.target.value as a way to obtain the previous value (see #3173 for more details). However, that is not the expected behavior and it wasn't documented as such.

File: src/js/components/Select/stories/typescript/Seasons.tsx
Comment (ShimiSun) [inline src/js/components/Select/stories/typescript/Seasons.tsx:20]:
```suggestion
// Type annotations can only be used in TypeScript files.
// Remove ': React.FC'  if you are not using TypeScript.
export const Seasons: React.FC = () => {
```

File: src/js/components/Select/stories/typescript/Seasons.tsx
Comment (ShimiSun) [inline src/js/components/Select/stories/typescript/Seasons.tsx:29]:
```suggestion
  // Type annotations can only be used in TypeScript files.
  // Remove ': string'  if you are not using TypeScript.
  const renderSeason = (season: string) => (
```

File: src/js/components/Select/stories/typescript/CreateOption.tsx
Comment (ShimiSun) [inline src/js/components/Select/stories/typescript/CreateOption.tsx:30]:
```suggestion
// Type annotations can only be used in TypeScript files.
// Remove ': string[]'  if you are not using TypeScript.
const defaultOptions: string[] = [];
```

**Review by ShimiSun — APPROVED** [review]
@ericsoderberghp I approve this PR, due to backward compatibility sensitivity I wanted another pair of eyes on this PR before merging.

File: src/js/components/Select/Select.js
Comment (jimenezff) [reply src/js/components/Select/Select.js:144]: @ShimiSun I think normally `event.target.value` is used to retrieve to the current value of the input element that changed, but for Grommet's Select component it's actually the previous value.

**Comment by jimenezff** [comment]
I added a deprecated tag in the latest commit https://github.com/grommet/grommet/pull/4694/commits/3129ae1f5b7a0f0ef11f8b2bee7d6c47f2cf5b42

File: src/js/components/Select/Select.js
Comment (ericsoderberghp) [reply src/js/components/Select/Select.js:144]: I think this approach is too narrowly focused, as the `event.target` will not be referring to the `<select />` DOM element but instead the `<button />` inside the drop. I think we should be using the technique we use in MaskedInput to first update the `<select />` value and then dispatching an event from it. See: https://github.com/grommet/grommet/blob/186b94413318dd7c4e30e9f8f47c124693acfd63/src/js/components/MaskedInput/MaskedInput.js#L238

Comment (ShimiSun) [reply src/js/components/Select/Select.js:144]: @jimenezff Any updates on this PR? It will be nice to get it on our next release.

**Comment by ShimiSun** [comment]
Hi @jimenezff Thanks for working on this fix! I'd like to get it in our next release, any chance we can get it to the finish line and incorporate the feedback?

File: src/js/components/Select/Select.js
Comment (jimenezff) [reply src/js/components/Select/Select.js:144]: @ShimiSun sorry for the delay. It looks like this may have already been solved with #5155. I'll check and revert my changes to `Select.js`

Comment (jimenezff) [reply src/js/components/Select/Select.js:144]: @ShimiSun I tried it but there are still issues with #3173. How about I remove the code related to that issue and we can just make this PR about type definition and the new TS story. If you agree, I can make that change today.

**Comment by jimenezff** [comment]
Hi @ShimiSun, regarding https://github.com/grommet/grommet/pull/4694#discussion_r617071599, would you prefer I create a new PR for Select's TS type definitions and story (I have the branch ready to go https://github.com/jimenezff/grommet/commit/cfd2702913e5fd2e880cca074a810b095e729cb1) or should I update this PR title and description?

**Comment by jcfilben** [comment]
Hi @jimenezff due to the age and merge conflicts on this PR lets close this one and create a new PR to update the type definitions

**Pull request closed without merging**
